### PR TITLE
doc: finish reworking the board porting guide

### PR DESCRIPTION
**Pull request by mbolivar**
Now that the west parts of this file [are in better shape](https://github.com/zephyrproject-rtos/zephyr/pull/23266), it seems a shame not to flesh out the rest a bit more.

It's been a while since we looked at this document, as it's still referring to boards (like Arduino 101) that are no longer supported by Zephyr, and is generally lacking in concrete, step-by-step advice for going from zero to working board.

Let's fix that.

### CODE REVIEW

**Comment by mbolivar** [comment]
@nashif I'm not intentionally ripping x86 out of the list of examples, I just don't know how your SoCs are named and I couldn't figure out the series/family relationships from existing boards docs. Any recommendation is welcome.

**Comment by mbolivar** [comment]
@hierophect FYI

**Comment by hierophect** [comment]
@mbolivar is there a good way to preview this form of documentation that you can recommend?

File: doc/guides/porting/board_porting.rst [hunk 1]
```diff
@@ -45,135 +31,333 @@ available in the source tree:
...
+hardware in the Devicetree Source (DTS) format (as usual, change ``plank`` to
+your board's name). If you're new to devicetree, see :ref:`devicetree-intro`.
+
+In general, :file:`plank.dts` should look like this:
+
+.. code-block:: none
+
+   /dts-v1/;
+   #include <your_soc_vendor/your_soc.dtsi>
+
```
Comment (hierophect) [inline doc/guides/porting/board_porting.rst]: Would it be reasonable to add a reference to including connector specific .dtsi files here, such as arduino_r3_connector.dtsi? In addition, is it generally not recommended to include other aliases for pins if they don't match a common connector? Several Micropython boards such as the Pyboard and PybNano do not use SoC pin names on their silkscreens.

File: doc/guides/porting/board_porting.rst [hunk 2]
```diff
@@ -45,135 +31,333 @@ available in the source tree:
...
+#. :file:`Kconfig.board`, :file:`Kconfig.defconfig`, :file:`plank_defconfig`:
+   software configuration in :ref:`kconfig` formats. This provides default
+   settings for software features and peripheral drivers.
+
+The optional files are:
+
+- :file:`board.cmake`: used for :ref:`flash-and-debug-support`. You can also
+  use this to build source code files whenever your board is used; see
+  :ref:`build_overview` for details.
+- :file:`index.rst`: documentation for your board. You only need this if you're
```
Comment (hierophect) [inline doc/guides/porting/board_porting.rst]: Perhaps you could consider adding a section that touches on the specifics of adding boards to the upstream? I think it'd be useful to set expectations for how much documentation Zephyr expects, and how to properly generate it - the Contribution Guidelines section has useful information about essential tools and standards like using Gitlint during a PR (might be good to link it) but documentation standards don't seem to be covered in an obvious spot. 

It'd be useful to know if you use a tool or IDE to generate the documentation to spec, with the correct line lengths, whitespace, etc.

Comment (hierophect) [reply doc/guides/porting/board_porting.rst]: I'd also add the directories in the image section for /img/image.png, since every board has at least one.

File: doc/guides/porting/board_porting.rst [hunk 3]
```diff
@@ -45,135 +31,333 @@ available in the source tree:
...
+  going to contribute the board to Zephyr.
+- :file:`plank.yaml`: a YAML file with miscellaneous metadata used by the
+  :ref:`sanitycheck_script`.
```
Comment (hierophect) [inline doc/guides/porting/board_porting.rst]: I notice you do not include pinmux.c here. I think a mention and a link to the appropriate documentation would be appropriate, since it is included in almost every ARM board.

File: doc/guides/porting/board_porting.rst [hunk 4]
```diff
@@ -45,135 +31,333 @@ available in the source tree:
...
+
+General recommendations
+***********************
+
+For consistency and to make it easier for users to build generic applications
+that are not board specific for your board, please follow these guidelines
+while porting.
+
+- Unless explicitly recommended otherwise by this section, leave peripherals
+  and their drivers disabled by default.
```
Comment (hierophect) [inline doc/guides/porting/board_porting.rst]: I'm not sure how, but more elaboration on this point might be useful? For a user looking at existing ARM implementations, most of them incorporate things like PWM, I2C, SPI and USB by default. If it is better to leave these disabled for the board definition, maybe it'd be useful to note the proper stage that they should be enabled in the development process?

File: doc/guides/porting/board_porting.rst (hunk 3 above)
Comment (mbolivar) [reply doc/guides/porting/board_porting.rst]: Well, it's not true on the boards my company makes, because we do our pinmux in DT the way the good Lord intended ;).

Your point is well taken, thank you. I had deliberately left this out because we're trying to get away from pinmux.c in general and replace everything with devicetree, but that doesn't work yet and that confuses people, so we might as well document the current state of the world.

File: doc/guides/porting/board_porting.rst (hunk 1 above)
Comment (mbolivar) [reply doc/guides/porting/board_porting.rst]: > Would it be reasonable to add a reference to including connector specific .dtsi files here, such as arduino_r3_connector.dtsi?

I can only find those files used in STM32 board definitions, so they appear to just be a convention followed for those boards. Nordic and NXP boards inline the connector node in the main DTS (e.g. nrf52840_pca10056.dts, frdm_k64f.dts).

Since there's not "one way to do it" in DTS, I'm just going to put the `connector` bit into the general recommendations instead:

```
- If your board uses a well-known connector standard (like Arduino, Mikrobus,
  Grove, or 96Boards connectors), add connector nodes to your DTS and configure
  pin muxes accordingly.
```

> In addition, is it generally not recommended to include other aliases for pins if they don't match a common connector?

Can you clarify what you'd like to see changed here? I'm having trouble parsing the double negative ("not recommended [...] if they don't"), sorry.

File: doc/guides/porting/board_porting.rst (hunk 2 above)
Comment (mbolivar) [reply doc/guides/porting/board_porting.rst]: > Perhaps you could consider adding a section that touches on the specifics of adding boards to the upstream?

Good idea. I've added one to the end.

> I think it'd be useful to set expectations for how much documentation Zephyr expects, and how to properly generate it - the Contribution Guidelines section has useful information about essential tools and standards like using Gitlint during a PR (might be good to link it) but documentation standards don't seem to be covered in an obvious spot.

A reference to doc/templates/board.tmpl is in the new section.

That file is the single point of truth for what constitutes minimal board documentation. Any requirements you think are missing from your experience getting a board upstreamed should be added to board.tmpl instead of board_porting.rst -- please LMK if you spot anything.

> It'd be useful to know if you use a tool or IDE to generate the documentation to spec,

A reference to the documentation generation page is in the new section.

> with the correct line lengths, whitespace, etc.

The SPOT for that is the linter scripts / their configuration, which change independently from this file, so I'd rather leave those types of details out of here so they don't go stale.

Thanks for your detailed review! I think I've now addressed all your comments, please take another look.

File: doc/guides/porting/board_porting.rst (hunk 4 above)
Comment (mbolivar) [reply doc/guides/porting/board_porting.rst]: > I'm not sure how, but more elaboration on this point might be useful?
>  For a user looking at existing ARM implementations, most of them incorporate things like PWM, I2C, SPI and USB by default.

Hmm, that seems to be covered below, no?

```
- Provide pin and driver configuration that matches the board's valuable
  components such as sensors, buttons or LEDs, and communication interfaces
  such as USB, Ethernet connector, or Bluetooth/Wi-Fi chip.

- If your board uses a well-known connector standard (like Arduino, Mikrobus,
  Grove, or 96Boards connectors), add connector nodes to your DTS and configure
  pin muxes accordingly.

- Configure components that enable the use of these pins, such as
  configuring an SPI instance to use the usual Arduino SPI pins.
```

> If it is better to leave these disabled for the board definition, maybe it'd be useful to note the proper stage that they should be enabled in the development process?

That is very, very use case specific, I think.

File: doc/guides/porting/board_porting.rst (hunk 2 above)
Comment (mbolivar) [reply doc/guides/porting/board_porting.rst]: > Thanks for your detailed review! I think I've now addressed all your comments, please take another look.

Whoops, forgot to do the pinmux updates. Now done.

File: doc/guides/porting/board_porting.rst (hunk 1 above)
Comment (hierophect) [reply doc/guides/porting/board_porting.rst]: For example, the Pyboard silkscreen uses Y1, Y2 -> Y12 and X1 -> X23 as pin labels. These are not following any existing convention, they're just pin nicknames. In this case, is it best practice to add a `.dtsi` file, or is it on the user to review the documentation of the board and use the actual SoC pin names in their code?

File: doc/guides/porting/board_porting.rst (hunk 4 above)
Comment (hierophect) [reply doc/guides/porting/board_porting.rst]: I think this could still be elaborated. In the Feather F405 instance, what's the cutoff for supporting a peripheral on a pin? Should PWM only be implemented when the pin has an LED, or even then? Should SPI and I2C be implemented on pins that reference them on the silkscreen? How about pins labeled as Analog pins (A0, A1, etc)? Or should peripherals ONLY be enabled for pins that aren't actually exposed to the user, such as for an on-chip accelerometer? 

Basically, what "counts" for stuff that should be configured at the board level?

**Comment by hierophect** [comment]
@mbolivar thanks for the edits. By the way, is there a better way of previewing these docs other than Github's preview that you could recommend?

File: doc/guides/porting/board_porting.rst (hunk 1 above)
Comment (mbolivar) [reply doc/guides/porting/board_porting.rst]: I'm not aware of any convention in devicetree where we define macros like Y2 or X23. I spot checked an STM32 board since this thread was originally about one, and didn't see anything there, either (the A0->A5 and D0->D15 silkscreen names are just in comments in the gpio-map property).

Am I missing something?

If you'd like to discuss adding macros or devicetree nodes/properties to make pin numbering in Zephyr more user friendly, @mnkp is the GPIO maintainer and probably the person to talk to. The mailing list or Slack #devicetree channel are a good place to do general discussion unless you want to open a separate issue here on GitHub..

I think feedback from folks like you at Adafruit would be valuable here.

**Comment by mbolivar** [comment]
> @mbolivar thanks for the edits. By the way, is there a better way of previewing these docs other than Github's preview that you could recommend?

You used to be able to load `https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/<PULL_REQUEST_NUMBER_GOES_HERE>/docs/guides/porting/board_porting.html` but it doesn't work anymore and I don't know why. Probably something about the move to GitHub actions.

File: doc/guides/porting/board_porting.rst (hunk 4 above)
Comment (mbolivar) [reply doc/guides/porting/board_porting.rst]: OK, I see.

I agree this could be improved, but answering this properly would start to get into the zoo of popular development board form factors and their conventions, which I'm not able to get into, so I will leave it to future work.

Might this be something Adafruit would like to help contributing?

File: doc/guides/porting/board_porting.rst (hunk 1 above)
Comment (hierophect) [reply doc/guides/porting/board_porting.rst]: I should stress I don't speak for Adafruit, this is all from a personal perspective. 

Let me try and rephrase the pin naming thing: is it possible to support custom aliases for pins in the case that the manufacturer has added nicknames for them that do not match their SoC names? I ask this because the pin naming on a board like the PyBoard is not a "connector" - it's just custom nicknames for pins that are unique to that board revision.

Comment (mbolivar) [reply doc/guides/porting/board_porting.rst]: > I should stress I don't speak for Adafruit, this is all from a personal perspective.

Sure!

> Let me try and rephrase the pin naming thing: is it possible to support custom aliases for pins in the case that the manufacturer has added nicknames for them that do not match their SoC names? I ask this because the pin naming on a board like the PyBoard is not a "connector" - it's just custom nicknames for pins that are unique to that board revision.

Do you really mean a devicetree alias? Then not to my knowledge, no. In devicetree, a gpio specifier is usually done with a phandle to the GPIO controller plus a specifier (usually a pin and some flags). It's not a node in the devicetree itself. Aliases can only point to nodes.

Comment (hierophect) [reply doc/guides/porting/board_porting.rst]: I'll work on some more trial projects and return to this if it seems like it's an obstacle when using hobby style development boards. Thanks for the input.

**Review by hierophect — APPROVED** [review]
Looks good to me, thanks for the edits and discussion.

**Review by jhedberg — APPROVED** [review]
